src/p4p/test/system_client_server.py

```diff
@@ -0,0 +1,631 @@
+        counter = q.get()
+        self.assertEqual(counter.counter.value, 5)
+        if self.MALCOLM:
+            # bitsets in pvaPy don't work, so it is everything at the moment
+            self.assertTrue(counter.asSet().issuperset({
+                "meta", "meta.fields", "counter", "zero"}))
+        else:
+            # TODO: timeStamp userTag didn't change value, but timeStamp did,
+            # should it be in the bitset?
+            self.assertEqual(counter.asSet(),
+                             {"counter.value", "counter.timeStamp",
+                              "counter.timeStamp.secondsPastEpoch"
+                              "counter.timeStamp.nanoseconds"})
+        self.ctxt.put("TESTCOUNTER.counter", 0, "value")
+        counter = q.get()
+        # TODO: Is this true even if you get a subset of changes?
+        self.assertStructureWithoutTsEqual(str(counter), str(counter_expected))
+
+    # Equivalent to:
+    #   pvget -m TESTCOUNTER -r meta.fields
+    def testMonitorSubfieldInitial(self):
+        q = Queue()
+        m = self.ctxt.monitor("TESTCOUNTER", q.put, "meta.fields")
+        self.addCleanup(m.close)
+        counter = q.get()
+        self.assertEqual(counter.getID(), "structure")
+        self.assertEqual(counter.asSet(), {"meta", "meta.fields"})
+        self.assertEqual(counter.meta.fields,
+                         ["health", "counter", "zero", "increment"])
+        fields_code = dict(counter.meta.type().aspy()[2])["fields"]
+        self.assertEqual(fields_code, "as")
+
+    # Equivalent to
+    #   pvget -m TESTCOUNTER.counter -r ""
+    def testMonitorDotted(self):
+        q = Queue()
+        m = self.ctxt.monitor("TESTCOUNTER.counter", q.put)
+        self.addCleanup(m.close)
+        counter = q.get()
+        self.assertEqual(counter.getID(), "epics:nt/NTScalar:1.0")
+        self.assertTrue(counter.asSet().issuperset({
+            "value", "alarm", "timeStamp"}))
+        self.ctxt.put("TESTCOUNTER.counter", 5, "value")
+        counter = q.get()
+        self.assertEqual(counter.value, 5)
+        if self.MALCOLM:
+            # bitsets in pvaPy don't work, so it is everything at the moment
+            self.assertTrue(counter.asSet().issuperset({
+                "value", "alarm", "timeStamp"}))
+        else:
+            # TODO: timeStamp userTag didn't change value, but timeStamp did,
+            # should it be in the bitset?
+            self.assertEqual(counter.asSet(),
+                             {"value", "timeStamp",
+                              "timeStamp.secondsPastEpoch"
+                              "timeStamp.nanoseconds"})
+        self.ctxt.put("TESTCOUNTER.counter", 0, "value")
+        counter = q.get()
+        self.assertEqual(counter.value, 0)
+
+    def testRpcRoot(self):
+        method = Value(Type([("method", "s")]), dict(method="increment"))
+        # TODO: Would be friendly to allow value=None to send empty structure
```

Comment:
Maybe.  It didn't occur to me as in the NTURI convention this is never an empty structure.